### PR TITLE
atanas/GROK-12164 Fix macromolecule detector for alphabet UN

### PR DESCRIPTION
**Pull request by tanas80**

### CODE REVIEW

File: packages/Bio/detectors.js [hunk 1]
```diff
@@ -55,6 +54,8 @@ class BioPackageDetectors extends DG.Package {
 
   RnaFastaAlphabet = new Set(['A', 'C', 'G', 'U']);
 
+  NumbersRawAlphabet = new Set(['0', '1', '2', '3', '4', '5', '6', '7', '8', '9']);
```
Comment (vdyma) [inline packages/Bio/detectors.js:57]: Generic variables should start with lowercase letter

File: packages/Bio/src/analysis/sequence-activity-cliffs.ts [hunk 2]
```diff
@@ -164,7 +166,7 @@ export function createLinesGrid(df: DG.DataFrame, colNames: string[]): DG.Grid {
     .init((i) => `${df.get(colNames[0], i)}#${df.get(colNames[1], i)}`);
   seqDiffCol.semType = 'MacromoleculeDifference';
```
Comment (LaykDimon) [inline packages/Bio/src/analysis/sequence-activity-cliffs.ts:167]: why don't you export the existing `enum SEM_TYPES` and use `SEM_TYPES.MACROMOLECULE_DIFFERENCE`?

File: packages/Bio/src/tests/detectors-tests.ts [hunk 3]
```diff
@@ -195,72 +207,72 @@ MWRSWY-CKHP
...
     };
   };
 
 
-  test('NegativeEmpty', async () => { await _testNeg(readCsv('csvDfEmpty', csvDfEmpty), 'col1'); });
-  test('Negative1', async () => { await _testNeg(readCsv('csvDf1', csvDf1), 'col1'); });
-  test('Negative2', async () => { await _testNeg(readCsv('csvDf2', csvDf2), 'col1'); });
-  test('Negative3', async () => { await _testNeg(readCsv('csvDf3', csvDf3), 'col1'); });
-  test('NegativeSmiles', async () => { await _testNeg(readCsv('csvDfSmiles', csvDfSmiles), 'col1'); });
+  test('NegativeEmpty', async () => { await _testNeg(readCsv(csvTests.negEmpty), 'col1'); });
```
Comment (LaykDimon) [inline packages/Bio/src/tests/detectors-tests.ts:221]: you use `'col1'` and `'seq'` multiple times, maybe it makes sense to put it into const?

File: packages/Bio/src/analysis/sequence-activity-cliffs.ts [hunk 4]
```diff
@@ -4,26 +4,28 @@ import * as DG from 'datagrok-api/dg';
...
+    const distance = stringArray[i] ? distanceMethod(stringArray[i], seq) : null;
     distances[i] = distance ? distance / Math.max((stringArray[i] as string).length, seq.length) : null;
   }
   return distances;
 }
 
-export async function getSimilaritiesMarix(dim: number, seqCol: DG.Column, df: DG.DataFrame, colName: string, simArr: DG.Column[])
-  : Promise<DG.Column[]> {
+export async function getSimilaritiesMatrix(
+  dim: number, seqCol: DG.Column, df: DG.DataFrame, colName: string, simArr: DG.Column[]
```
Comment (vdyma) [inline packages/Bio/src/analysis/sequence-activity-cliffs.ts:27]: Columns should have types

File: packages/Bio/src/utils/cell-renderer.ts [hunk 5]
```diff
@@ -118,20 +132,29 @@ export class MacromoleculeSequenceCellRenderer extends DG.GridCellRenderer {
...
+    const referenceSequence: string[] = splitterFunc(
+      ((tempReferenceSequence != null) && (tempReferenceSequence != '')) ?
+        tempReferenceSequence : tempCurrentWord ?? '');
+    const monomerWidth: string = (tempMonomerWidth != null) ? tempMonomerWidth : 'short';
 
+    let gapRenderer = 5;
+    let maxIndex = 0;
     let maxLengthOfMonomer = 8;
 
     if (monomerWidth === 'short') {
```
Comment (LaykDimon) [inline packages/Bio/src/utils/cell-renderer.ts:160]: magic const?

File: packages/Bio/src/utils/convert.ts [hunk 6]
```diff
@@ -87,6 +87,11 @@ export async function convertDo(
   const newColumn = converter.convert(targetNotation, separator);
   srcCol.dataFrame.columns.add(newColumn);
 
+  // Call detector directly to escape some error on detectSemanticTypes
+  const semType = await grok.functions.call('Bio:detectMacromolecule', {col: newColumn});
```
Comment (StLeonidas) [inline packages/Bio/src/utils/convert.ts:91]: Is this the same package?

File: packages/Bio/detectors.js [hunk 7]
```diff
@@ -109,7 +110,8 @@ class BioPackageDetectors extends DG.Package {
       }
 
       const decoyAlphabets = [
-        ['SMILES', this.SmilesRawAlphabet, 0.30],
+        ['NUMBERS', this.NumbersRawAlphabet, 0.25],
+        ['SMILES', this.SmilesRawAlphabet, 0.25],
```
Comment (StLeonidas) [inline packages/Bio/detectors.js:114]: What are these numbers?

File: packages/Bio/detectors.js [hunk 8]
```diff
@@ -306,28 +325,27 @@ class BioPackageDetectors extends DG.Package {
...
+
     return prod;
   }
 
   /** For trivial checks split by single chars*/
   getSplitterAsChars(lengthLimit) {
     return function(seq) {
       return seq.split('', lengthLimit);
-    }.bind(this);
+    };
```
Comment (StLeonidas) [inline packages/Bio/detectors.js:348]: Do you need one line function?

File: packages/Bio/src/utils/cell-renderer.ts [hunk 9]
```diff
@@ -149,43 +172,41 @@ export class MacromoleculeSequenceCellRenderer extends DG.GridCellRenderer {
...
           if (textSize > (maxLengthWords[index] ?? 0))
             maxLengthWords[index] = textSize;
-          if (index > maxIndex) {
-            maxIndex = index;
-          }
+          if (index > maxIndex) maxIndex = index;
         });
         samples += 1;
       }
       const minLength = 3 * 7;
```
Comment (StLeonidas) [inline packages/Bio/src/utils/cell-renderer.ts:179]: What 3 and 7 stands for?

File: packages/Bio/src/utils/cell-renderer.ts (hunk 5 above)
Comment (tanas80) [reply packages/Bio/src/utils/cell-renderer.ts:160]: Create ticket

Comment (tanas80) [reply packages/Bio/src/utils/cell-renderer.ts:160]: [GROK-12407](https://reddata.atlassian.net/browse/GROK-12407)
 Bio | Tools | Macromolecule cell renderer refactoring

File: packages/Bio/detectors.js (hunk 7 above)
Comment (tanas80) [reply packages/Bio/detectors.js:114]: These values are chosen empirically to correctly classify according to the test data samples.

**Comment by vdyma** [comment]
How come there are so many commits and file changes? This PR is too broad now and doesn't fit to the corresponding branch name and task.

**Pull request merged**